tests/sycl/relational.cpp

```diff
@@ -0,0 +1,237 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "sycl_test_suite.hpp"
+
+#include <boost/mpl/joint_view.hpp>
+
+#include <cmath>
+#include <cfloat>
+
+BOOST_FIXTURE_TEST_SUITE(rel_tests, reset_device_fixture)
+
+// list of types classified as "genfloat" in the SYCL standard
+using rel_test_genfloats = boost::mpl::list<
+  float,
+  // vec<T,1> is not genfloat according to SYCL 2020. It's unclear
+  // if this is an oversight or intentional.
+  //cl::sycl::vec<float, 1>
+  cl::sycl::vec<float, 2>,
+  cl::sycl::vec<float, 3>,
+  cl::sycl::vec<float, 4>,
+  cl::sycl::vec<float, 8>,
+  cl::sycl::vec<float, 16>,
+  double,
+  //cl::sycl::vec<double, 1>,
+  cl::sycl::vec<double, 2>,
+  cl::sycl::vec<double, 3>,
+  cl::sycl::vec<double, 4>,
+  cl::sycl::vec<double, 8>,
+  cl::sycl::vec<double, 16>>;
+
+
+
+namespace {
+
+  template<typename DT, int N>
+  using vec = cl::sycl::vec<DT, N>;
+
+  // utility type traits for generic testing
+
+  template<typename T>
+  struct vector_length {
+    static constexpr int value = 0;
+  };
+  template<typename DT, int N>
+  struct vector_length<vec<DT, N>> {
+    static constexpr int value = N;
+  };
+  template<typename T>
+  constexpr int vector_length_v = vector_length<T>::value;
+
+  template<typename T>
+  struct vector_elem {
+    using type = T;
+  };
+  template<typename DT, int D>
+  struct vector_elem<vec<DT, D>> {
+    using type = DT;
+  };
+  template<typename T>
+  using vector_elem_t = typename vector_elem<T>::type;
+
+  // utility functions for generic testing
+
+  template<typename DT, int D, std::enable_if_t<D<=4, int> = 0>
+  auto get_rel_input(cl::sycl::vec<DT, 16> v) {
+    return std::get<D>(std::make_tuple(
+      v.s0(),
+      vec<DT, 1>(v.s0()),
+      vec<DT, 2>(v.s0(), v.s1()),
+      vec<DT, 3>(v.s0(), v.s1(), v.s2()),
+      vec<DT, 4>(v.s0(), v.s1(), v.s2(), v.s3())));
+  }
+  template<typename DT, int D, std::enable_if_t<D==8, int> = 0>
+  auto get_rel_input(cl::sycl::vec<DT, 16> v) {
+    return vec<DT, 8>(v.s0(), v.s1(), v.s2(), v.s3(), v.s4(), v.s5(), v.s6(), v.s7());
+  }
+  template<typename DT, int D, std::enable_if_t<D==16, int> = 0>
+  auto get_rel_input(cl::sycl::vec<DT, 16> v) {
+    return v;
```

Comment:
This is just to obtain n-component subvector of a 16-component vector, right? Couldn't you just do something like:
```c++
template<class T, int N>
auto get_subvector(vec<T, 16> v) {
  if constexpr(N==1) {
    return vec<T,1>{v.swizzle<0>()};
  else if constexpr(N==2){
    return vec<T,2>{v.swizzle<0,1>()};
  } else if constexpr(N==3){
    ....
  } ....
}
```

Reply:
Done, but nvc++ doesn't like it, we need to think of a workaround...